src/libraries/System.Private.CoreLib/src/Internal/Console.Windows.cs

```diff
@@ -8,16 +8,55 @@ namespace Internal
 {
     public static partial class Console
     {
-        private static readonly IntPtr s_outputHandle =
-            Interop.Kernel32.GetStdHandle(Interop.Kernel32.HandleTypes.STD_OUTPUT_HANDLE);
+        public static void Write(string s)
+        {
+            WriteCore(Interop.Kernel32.GetStdHandle(Interop.Kernel32.HandleTypes.STD_OUTPUT_HANDLE), s);
+        }
 
-        public static unsafe void Write(string s)
+        public static partial class Error
         {
-            byte[] bytes = Encoding.UTF8.GetBytes(s);
-            fixed (byte* pBytes = bytes)
+            public static void Write(string s)
             {
-                Interop.Kernel32.WriteFile(s_outputHandle, pBytes, bytes.Length, out _, IntPtr.Zero);
+                WriteCore(Interop.Kernel32.GetStdHandle(Interop.Kernel32.HandleTypes.STD_ERROR_HANDLE), s);
             }
         }
+
+        private static Encoding s_encoding;
+
+        private static unsafe void WriteCore(IntPtr handle, string s)
+        {
+            if (IsHandleRedirected(handle))
+            {
+                if (s_encoding == null)
+                {
+                    s_encoding =
+                        EncodingHelper.GetSupportedConsoleEncoding((int)Interop.Kernel32.GetConsoleOutputCP());
+                }
+
+                byte[] bytes = s_encoding.GetBytes(s);
+                fixed (byte* pBytes = bytes)
+                {
+                    Interop.Kernel32.WriteFile(handle, pBytes, bytes.Length, out _, IntPtr.Zero);
+                }
+            }
+            else
+            {
+                fixed (char* pChars = s)
+                {
+                    Interop.Kernel32.WriteConsole(handle, (byte*)pChars, s.Length, out _, IntPtr.Zero);
+                }
+            }
+        }
+
+        private static bool IsHandleRedirected(IntPtr handle)
```

Comment:
The CoreCLR unmanaged equivalent (`PrintToStdErr` - used by unmanaged failfast) does not have any of this special handling for redirected vs. non-redirected output.